File: src/tester/categorizer.ts
export class ModelCategorizer {
  private rules: Record<string, { keywords: string[]; exclude_keywords?: string[]; description?: string }>;

  constructor(config: any) {
    this.rules = config.model_categories || {};
  }

  private classify(modelId: string): string {
    const mid = modelId.toLowerCase();
    const categories = Object.keys(this.rules);

    for (const [cat, rule] of Object.entries(this.rules)) {
      const keywords = rule.keywords || [];
      const excludeKeywords = rule.exclude_keywords || [];

      // If keywords list is empty, it should not match directly (it's the fallback)
      if (keywords.length === 0) {
        continue;
      }

      const matchesKeyword = keywords.some((kw) => mid.includes(kw.toLowerCase()));
      const matchesExclude = excludeKeywords.some((ex) => mid.includes(ex.toLowerCase()));

      if (matchesKeyword && !matchesExclude) {
        return cat;
      }
    }

    return (categories.length > 0 ? categories[categories.length - 1] : "general_chat") ?? "general_chat";
  }

  categorize(models: any[]): Record<string, any[]> {
    const groups: Record<string, any[]> = {};
    for (const cat of Object.keys(this.rules)) {
      groups[cat] = [];
    }

    for (const model of models) {
      const modelId = model.id || "";
      const cat = this.classify(modelId);
      if (!groups[cat]) {
        groups[cat] = [];
      }
      groups[cat].push(model);
    }

    const ordered: Record<string, any[]> = {};
    for (const cat of Object.keys(this.rules)) {
      if (groups[cat] && groups[cat].length > 0) {
        ordered[cat] = groups[cat];
      }
    }
    return ordered;
  }
}

File: python_legacy/tester/categorizer.py
# -*- coding: utf-8 -*-
"""模型分类模块：基于关键词匹配将模型分组"""

import logging
from collections import defaultdict
from typing import Any, Dict, List

logger = logging.getLogger(__name__)


class ModelCategorizer:
    """基于模型 ID 关键词进行智能分组"""

    def __init__(self, config: dict):
        self.rules = config["model_categories"]  # 保持 YAML 中定义的顺序

    def _classify(self, model_id: str) -> str:
        """返回模型所属分类 key"""
        mid = model_id.lower()
        for cat, rule in self.rules.items():
            keywords = rule.get("keywords", [])
            if any(kw in mid for kw in keywords):
                return cat
        # O-4: 兜底分类取 config 中最后一个分类（而非硬编码 general_chat）
        return list(self.rules.keys())[-1] if self.rules else "general_chat"

    def categorize(
        self, models: List[Dict[str, Any]]
    ) -> Dict[str, List[Dict[str, Any]]]:
        """对模型列表进行分组，返回 {category_key: [model_dict, ...]}"""
        groups: Dict[str, List] = defaultdict(list)
        for model in models:
            model_id = model.get("id", "")
            cat = self._classify(model_id)
            groups[cat].append(model)
            logger.debug(f"  [{cat}] {model_id}")

        # 只返回有数据的分组，且按 YAML 定义顺序
        ordered = {}
        for cat in self.rules:
            if groups[cat]:
                ordered[cat] = groups[cat]
        return ordered
